Add atRoot parameter to `Value.DeepRemove`

turbolent opened the pull request:
Work towards #2882, in particular https://github.com/onflow/cadence/pull/2882/commits/593ceada94c96e4a6361c73f56fd3ad42b6ae6de

## Description

- Allow `Value.DeepRemove` implementations to differentiate when a call is at the root level, or if it is a recursive call, by adding a parameter `atRoot bool`.
- Only validate the atree value of interpreter values backed by atree values (`CompositeValue`, `DictionaryValue`, and `ArrayValue`) at the root level

______

<!-- Complete: -->

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

fxamacker reviewed (approved):
Thanks for doing this! :+1:   Looks good to me, and I only have one comment.

`maybeValidateAtreeValue()` has two checks: atree value validation and storage slab health.

Since `atRoot` only affects storage slab health check, maybe we can add parameter to `maybeValidateAtreeValue()` to skip storage health check if `atRoot` is false.

turbolent commented:
@fxamacker Good idea! Split the validation into two functions, and changed call-sites to always validate the atree value, and only conditionally validate atree storage in a81c720e141820d292d48a32a02c7ef78cecf7c0. Could you please have another look?

fxamacker reviewed (approved):
LGTM!  Thanks for splitting the checks into two functions!
